### PR TITLE
Adding a Character Chat Agent

### PR DESCRIPTION
**Pull request by treppers**
# Character Chat Agent

Fork of the conversational chat agent, but added in the concept of personality and fixed some issues I was getting with the output_parser.  You can provide a {character_summary} that will give your bot that personality.

I was noticing some occasional weird responses with the conversation_chat agent so rewrote the entire prompt which seemed to have fixed some of them.  Also noticed some errors with output_parser (but will submit that as a separate PR).

Changes with the prompt:

- Moving all the TOOLS and FORMAT INSTRUCTIONS into the first System Message. 
- Keeping the actual Human Message to include the original input message.
- Introduce a third voice called TOOL, that only speaks via the System Message
- Created a More Instructions section instead of User Input to try to trigger the Final Answer
- Also made sure to remove most of the you's and i's and replace it with nouns (Assistant, Human, TOOL)

Output_parser changes:
- The agent sometimes doesn't want to respond with ````json` but instead will just say ````` which breaks the split function.
- The agent sometimes doesn't doesn't want to follow the instructions and doesn't respond with any markdown (this is usually when the prompt contains enough information to do the completion).  To fix this I put a try-except to catch the when JSON fails.
- The agent sometimes runs out of token and doesn't close the markdown snippet.  Created a case to handle this case.

Community members can review the PR once tests pass. Tag maintainers/contributors who might be interested:

@hwchase17

### CODE REVIEW

File: langchain/agents/character_chat/prompt.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+TEMPLATE_TOOL_RESPONSE = """TOOL RESPONSE: 
+---------------------
+{observation}
+
+MORE INSTRUCTIONS
+--------------------
+Given the entire TOOLS RESPONSE, 
+- If the USER'S ORIGINAL INPUT isn't answered using ONLY the information obtained by TOOL, try the same tool with a different input or another tool.
+- Otherwise, how would Assistant respond using information obtained from TOOL, Assistant must NOT mention the tool or tool names - the user has no context of any TOOL RESPONSES!
```
Comment (mbchang) [inline langchain/agents/character_chat/prompt.py:108]: It looks like this is a run-on sentence; could this be re-written? Perhaps something like:
```
Otherwise, to respond using information obtained from TOOL, Assistant must NOT mention the tool or tool names - the user has no context of any TOOL RESPONSES!
```

File: langchain/agents/character_chat/__init__.py [hunk 2]
```diff
@@ -0,0 +1 @@
+"""An agent designed to hold a conversation with a character in addition to using tools."""
```
Comment (mbchang) [inline langchain/agents/character_chat/__init__.py:1]: Would it be more accurate to rephrase this as "An agent designed to hold a conversation as a given character in addition to using tools."

File: langchain/agents/character_chat/base.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
+"""An agent designed to hold a conversation with a character in addition to using tools."""
```
Comment (mbchang) [inline langchain/agents/character_chat/base.py:1]: Would it be more accurate to rephrase this as "An agent designed to hold a conversation as a given character in addition to using tools."

File: langchain/agents/character_chat/base.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+                                    HumanMessagePromptTemplate,
+                                    MessagesPlaceholder,
+                                    SystemMessagePromptTemplate)
+from langchain.schema import (AgentAction, AIMessage, BaseMessage,
+                              BaseOutputParser, SystemMessage)
+from langchain.tools.base import BaseTool
+
+
+class CharacterChatAgent(Agent):
+    """An agent designed to hold a conversation in addition to using tools."""
```
Comment (mbchang) [inline langchain/agents/character_chat/base.py:28]: Would it be more accurate to rephrase this as "An agent designed to hold a conversation as a given character in addition to using tools."

File: langchain/agents/character_chat/prompt.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+Assistant can ask the TOOL to use tools to look up information that may be helpful in answering the users original question. The tools the TOOL can use are:
+
+{tools}
+
+RESPONSE FORMAT INSTRUCTIONS
+----------------------------
+
+When responding to the TOOL, please output a response in one of two formats:
+
+**Option 1:**
```
Comment (mbchang) [inline langchain/agents/character_chat/prompt.py:79]: `**Option #1:**` instead of `**Option 1:**`

File: langchain/agents/character_chat/prompt.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+}}}}
+```"""
+
+TEMPLATE_TOOL_RESPONSE = """TOOL RESPONSE: 
+---------------------
+{observation}
+
+MORE INSTRUCTIONS
+--------------------
+Given the entire TOOLS RESPONSE,
```
Comment (mbchang) [inline langchain/agents/character_chat/prompt.py:106]: `TOOLS` -> `TOOL's`

File: langchain/agents/character_chat/prompt.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+TOOLS
+------
+Assistant can ask the TOOL to use tools to look up information that may be helpful in answering the users original question. The tools the TOOL can use are:
+
+{tools}
+
+RESPONSE FORMAT INSTRUCTIONS
+----------------------------
+
+When responding to the TOOL, please output a response in one of two formats:
```
Comment (mbchang) [inline langchain/agents/character_chat/prompt.py:77]: `When responding to the TOOL` -> `When responding to TOOL`

**Comment by mbchang** [comment]
Thanks @treppers for the PR! This kind of agent looks like it could make writing dialogue agents much more streamlined.
- It looks like you have made a few changes to the system prompt. Would you mind sharing some concrete examples (as a  jupyter notebook or code snippet) that illustrate the tangible differences between using the original prompt scheme vs your proposed prompt scheme? Such an ablation study would be helpful to know whether your proposed changes are specific to the CharacterChatAgent or whether they should be incorporated more broadly into other ConversationAgents as well. Specifically, for your following changes, do you have examples of what the response would look like with and without the change? 
  - Moving all the TOOLS and FORMAT INSTRUCTIONS into the first System Message.
  - Keeping the actual Human Message to include the original input message.
  - Introduce a third voice called TOOL, that only speaks via the System Message
  - Created a More Instructions section instead of User Input to try to trigger the Final Answer
  - Also made sure to remove most of the you's and i's and replace it with nouns (Assistant, Human, TOOL)
- The CharacterChatAgent could be generalized to talk to not just the human but other characters. Could we replace "human" and "user" with "conversation partner" in the system prompt?

File: langchain/agents/character_chat/__init__.py (hunk 2 above)
Comment (treppers) [reply langchain/agents/character_chat/__init__.py:1]: Definitely correct!

File: langchain/agents/character_chat/base.py (hunk 4 above)
Comment (treppers) [reply langchain/agents/character_chat/base.py:28]: Same comment as above.  Incorporating.

File: langchain/agents/character_chat/prompt.py (hunk 5 above)
Comment (treppers) [reply langchain/agents/character_chat/prompt.py:79]: Adding this into the prompt.

**Comment by treppers** [comment]
Trying to put together a notebook that will do a good compare and contrast.

That should be inbound.

I do think the Conversational Chat agent needs updating and so when I get the notebook up you can see the issues I found.

Definitely open to expanding the prompt to apply to conversation partner.  Would just need to test it to make sure it doesn't think the TOOL voice is a conversation partner.

Updating this branch with a notebook to execute this agent along with the suggested tweaks.  Also found a bug in the prompt that was causing the Option #1 response to include additional conversational responses in addition to the markdown.

**Comment by mbchang** [comment]
Thanks for the update @treppers! Curious to see what you find in the compare-and-contrast.

**Comment by efriis** [comment]
Hey @mbchang ! This PR no longer lines up with the current directory structure of the library (would need to be in /libs/langchain/langchain instead of /langchain). Would you be interested in updating to the new format, or would it be better to close this and someone can work on a new PR off the current state of the library (using this as a reference implementation)?

**Comment by efriis** [comment]
I'll close it for now, and let me know if you'd like to reopen!
